doc/elements-tx-format.md

```diff
@@ -0,0 +1,562 @@
+
+#### Variable Length Integer (VarInt)
+
+This data type is derived from Bitcoin and allows an integer to be encoded depending on the represented value, to save space. Variable length integers always precede an vector of a type of data that may vary in length. Longer numbers are encoded in little-endian.
+
+Refer to the [Bitcoin protocol documentation](https://en.bitcoin.it/wiki/Protocol_documentation#Variable_length_integer) for a full description on how this type is serialized.
+
+#### Vector\<Type\>
+
+Each `Vector` begins with a `VarInt` describing the number of items it contains.
+
+If the vector is of type `hex`, then the size / structure of each individual item is not known in advance. In this case, each item begins with a `VarInt` describing its size `s` in bytes, followed by `s` bytes which should be interpreted as the item itself.
+Otherwise, size prefixes are omitted, and each item should be interpreted in accordance with the vector's type.
+
+In other words, the vector is serialized as follows: `[Length (n)][Item #1][Item #2][...][Item #n]`.
+
+Refer to the examples section below for more concrete examples of serialized vectors.
+
+#### TxIn
+
+|Field|Required|Size|Data Type|Encoding|Notes| 
+|-----|--------|-----|---------|--------|-----|
+| Output Hash (TXID) | Yes | 32 bytes | `hex` | [^1] | |
+| Output Index | Yes | 4 bytes | `uint32_t` | Little-endian | **Input is a coinbase**: `0xffffffff`<br><br>The two most significant bits are reserved for flags.<br><br>**Input is a pegin:** second most significant bit is 1.<br><br>**Input has an asset issuance:** most significant bit is 1. |
+| ScriptSig Length | Yes | Varies | `VarInt` | | Set to `0x00` if the transaction is SegWit and the witness contains the signature. |
+| ScriptSig | If ScriptSig Length is non-zero | Varies | `hex` | Big-endian | |
+| Sequence | Yes | 4 bytes | `uint32_t` | Little-endian | |
+| Asset Issuance | Only if the transaction has an issuance (as indicated by the Output Index) | Varies | `AssetIssuance` |  | |
+
+[^1]: The hex encodings of hashes by the Elements client (TXID, asset ID) are byte-reversed, and so the bytes will need to be re-reversed to match the serialized data. This is the same situation as in Bitcoin. For example, the hash `1123...deff` would be displayed by the Bitcoin and Elements clients as `ffde...2311`. This is primarily for historical reasons: the Bitcoin client has always interpreted and displayed hashes as little-endian integers and parsed their bytes in reverse order.
+
+#### TxOut
+
+|Field|Required|Size|Data Type|Encoding|Notes|
+|-----|--------|-----|---------|--------|-----|
+| Asset | Yes | 33 bytes | `ConfidentialAsset` | | Cannot be null. |
+| Amount | Yes | 9 or 33 bytes | `ConfidentialAmount` | | Cannot be null. |
+| Nonce | Yes | 1 or 33 bytes | `ConfidentialNonce` | | |
+| ScriptPubKey Length | Yes | Varies | `VarInt` | | | |
+| ScriptPubKey | If ScriptPubKey Length is non-zero | Varies | `hex` | Big-endian | |
+
```

Comment:
Do we want a section on the definition and requirements of "UNSPENDABLE" outputs?

Reply:
I would suggest not, it seems out of scope.

Reply:
There are restrictions on outputs with explicit 0 / rangeproofs with 0 and the UNSPENDABLE pubkey condition.

Reply:
Right, but I consider those similar to "there are restrictions on inputs which have bad signatures" ... i.e.this is a validation concern, not a parsing one.

Reply:
I disagree about it not being a parsing issue, but I'll accept it.

Reply:
I have two arguments for it being a validation issue:

* In order to check this you'd have to call into secp-zkp to extract parameters from the rangeproof itself, which otherwise parsers can simply treat as a byte-vector. So if this were a parsing issue then all parsers would need secp-zkp and to "understand the crypto" when currently they don't
* Post-bulletproofs the minimum value will be implicit, so it definitely won't be a parser concern then

Reply:
That is a pretty strong argument.